### PR TITLE
Forward solver adapted for automatic differentiation.

### PR DESCRIPTION
**Pull request by Ig-dolci**
This PR seeks to update the acoustic forward and FWI code using the automated adjoint solver. 

New tools:

- Two-level parallelism: Mesh and source with `firedrake.adjoint.EnsembleReducedFunctional`
- Enable checkpointing.

What is missing: a proper absorbing boundary condition.

### CODE REVIEW

File: demos/with_automatic_differentiation/run_fwi_ad.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        c, compute_functional=False, true_data_receivers=None, annotate=False
+):
+    if annotate:
+        fire_ad.continue_annotation()
+    if model["parallelism"]["type"] is None:
+        outfile = fire.VTKFile("solution.pvd")
+        receiver_data = []
+        J = 0.0
+        for sn in range(len(model["acquisition"]["source_pos"])):
+            rec_data, J_val = forward_solver.execute(c, sn, wavelet)
```
Comment (Ig-dolci) [inline demos/with_automatic_differentiation/run_fwi_ad.py:94]:
```suggestion
            rec_data, J_val = forward_solver.execute_acoustic(c, sn, wavelet)
```

**Comment by Ig-dolci** [comment]
Spyro needs to update the firedrake version to have the checkpointing test pass.

**Comment by Olender** [comment]
> Spyro needs to update the firedrake version to have the checkpointing test pass.

I can install the latest version on the runner. We're currently using the version from 2024-07-19. Please let me know which version we should target for this update.

**Comment by Ig-dolci** [comment]
> > Spyro needs to update the firedrake version to have the checkpointing test pass.
> 
> I can install the latest version on the runner. We're currently using the version from 2024-07-19. Please let me know which version we should target for this update.

The latest.

**Comment by Olender** [comment]
Ok, I have already installed everything in the runner and changed the firedrake location in the .yml workflow file

**Review by Olender — APPROVED** [review]

**Pull request merged**